libfoundation/src/foundation-objc.mm

```diff
@@ -383,6 +383,794 @@ MC_DLLEXPORT_DEF uintptr_t MCObjcObjectGetActionProxySelector(void)
+                                               p_context,
+                                               t_protocol))
+            return false;
+    }
+    
+    com_livecode_MCObjcFormalDelegate *t_proxy =
+    [[com_livecode_MCObjcFormalDelegate alloc] initForProtocol: t_protocol withHandlerMapping:p_handler_mapping withContext:p_context];
+    
+    MCObjcObjectRef t_proxy_object = nullptr;
+    if (!MCObjcObjectCreateWithRetainedId(t_proxy, t_proxy_object))
+    {
+        [t_proxy release];
+        return false;
+    }
+    
+    r_object = t_proxy_object;
+    return true;
+}
+
+extern "C" MC_DLLEXPORT_DEF
+bool MCObjcCreateDelegate(MCStringRef p_protocol_name, MCArrayRef p_handler_mapping, MCObjcObjectRef& r_object)
+{
+    return MCObjcCreateDelegateWithContext(p_protocol_name,
+                                           p_handler_mapping,
+                                           nullptr,
+                                           r_object);
+}
+
+extern "C" MC_DLLEXPORT_DEF
+bool MCObjcCreateInformalDelegateWithContext(MCProperListRef p_foreign_handlers, MCArrayRef p_handler_mapping, MCValueRef p_context, MCObjcObjectRef& r_object)
+{
+    MCAutoArrayRef t_selector_mapping;
+    if (!MCArrayCreateMutable(&t_selector_mapping))
+        return false;
+    
+    for (uindex_t i = 0; i < MCProperListGetLength(p_foreign_handlers); ++i)
+    {
+        MCValueRef t_element =
+            MCProperListFetchElementAtIndex(p_foreign_handlers, i);
+        
+        if (MCValueGetTypeCode(t_element) != kMCValueTypeCodeHandler)
+            return false;
+        
+        MCHandlerRef t_handler = static_cast<MCHandlerRef>(t_element);
+        
+        const MCHandlerCallbacks *t_callbacks = MCHandlerGetCallbacks(t_handler);
+        if (t_callbacks->query == nullptr)
+            return false;
+        
+        SEL t_selector = nullptr;
+        if (!t_callbacks->query(MCHandlerGetContext(t_handler),
+                                kMCHandlerQueryTypeObjcSelector,
+                                &t_selector))
+            return false;
+        
+        if (t_selector == nullptr)
+            return false;
+        
+        NSString *t_sel_string = NSStringFromSelector(t_selector);
+        MCAutoStringRef t_sel_stringref;
+        if (!MCStringCreateWithCFStringRef((CFStringRef)t_sel_string, &t_sel_stringref))
+            return false;
+        
+        MCNewAutoNameRef t_key;
+        if (!MCNameCreate(*t_sel_stringref, &t_key))
+            return false;
+        
+        if (!MCArrayStoreValue(*t_selector_mapping, false, *t_key, t_element))
+            return false;
+    }
+
+    MCAutoArrayRef t_selector_mapping_final;
+    if (!MCArrayCopy(*t_selector_mapping, &t_selector_mapping_final))
+        return false;
+    
+    MCNameRef t_key;
+    MCValueRef t_value;
+    uintptr_t t_iterator = 0;
+    while (MCArrayIterate(p_handler_mapping, t_iterator, t_key, t_value))
```

Comment:
@livecodeali I think it would be more helpful to do this the other way around and use `protocol_copyMethodDescriptionList` to ensure that the handler mapping conforms to the required methods otherwise things will crash because required methods won't be guarded by `respondsToSelector` and we might forget to add one to the array. Perhaps it should check both ways?

Reply:
Hmm, I hadn't thought about required ones in that way - good point